### ISSUE TITLE
jQuery plugin namespace

### ISSUE THREAD

**Issue by yairEO** [gilmoreorless/jquery-dom-line#4]
$.line = function (from, to, options) {

The name you gave it, `line` is extremely improper because it is too general. you should use a more specific name so it won't get overwritten accidentally. it's best practice.
**also**, authoring jQuery plugins is by using the `$.fn` prototype shortcut. read more here:
http://stackoverflow.com/a/4083362/104380

## My recommendation is to change it to:

    $.fn.drawLine = function (from, to, options) {

also, the defaults should be in a variable and not on the namespace itself, in my opinion.
the user can already pass new options anyway to the plugin, which extend yours.

## Another thing:
it makes no sense, at all, to make this a jQuery plugin. it should be a simple javascript constructor, not a jQuery plugin.

a jQuery plugin is something you use on elements, like this:

    from.drawLine(to); // where "from" is a jQuery element and to can be DOM node or jQuery element

And not like:

    $.line(from, to)

**Comment by yairEO** [comment]
you know what, I'll just fork and re-write it and show you

**Comment by gilmoreorless** [comment]
Hi there, thanks for the interest in this old project. To be honest, I'd kind of forgotten about the existence of it, but I'll answer your points about the code style as best as I can remember.

1. Attaching a jQuery plugin to `$.fn` only applies to methods that directly operate on elements, so that you can call `$(element).thing()`. This plugin does not operate _on_ any element directly, unless one is referenced in the options – it only operates on x/y point coordinates. Therefore it belongs on the top-level `$` object, along with other general utility methods such as `$.ajax()`, `$.param()`, `$.contains()`, `$.extend()` and more. Redefining the plugin to be `$.fn.drawLine()` would result in every jQuery element wrapper having a method called `drawLine()` that is completely unrelated to that element.
I do agree with your point on the name `.line()` being too generic, though.

2. Attaching defaults to the function definition was a common pattern for jQuery plugins at the time this was written, and is still recommended on the [learn.jquery.com](https://learn.jquery.com/plugins/advanced-plugin-concepts/) site:
> An improvement we can, and should, make to the code above is to expose the default plugin settings. This is important because it makes it very easy for plugin users to override/customize the plugin with minimal code.

3. Your suggested API of `from.drawLine(to)` assumes that `from` and `to` are elements, which is not what this plugin does at all. The `from` and `to` arguments of this plugin are basic objects containing `x` and `y` properties, not elements. This is the same reason the plugin is not defined on the `$.fn` namespace.

4. Regarding this utility being a jQuery plugin rather than vanilla JS. Keep in mind it was first written 6 years ago (and has had no updates in nearly 4 years). Back then, browser support for CSS transforms was inconsistent, and jQuery attribute hooks provided the simplest way to smooth over the differences. This plugin simply built on top of the [jquery.transform.js](http://louisremi.github.io/jquery.transform.js/) plugin. If I were writing it from scratch today I certainly wouldn't write it as a jQuery plugin. Then again, I probably wouldn't write it at all, because...

5. This plugin started as a quick experiment that I needed for the examples of a different plugin. At that time, IE7 and 8 were still prevalent and did not support SVG or Canvas. Drawing a straight line using a standard DOM element seemed like a great way to cheat and work around it (of course, I still needed the transform plugin to support those old IEs). Looking back, it was an idea that wasn't exactly terrible, but wasn't great either. I should have just used SVG or Canvas with a VML fallback. In today's browsers there's not really a need for a plugin like this; there are much better tools available.

In summary, I appreciate your attention to detail, but I actually think it's better to put this plugin to rest, and take advantage of the far better (and more flexible) drawing APIs that are readily available today.
